src/communicator.h

```diff
@@ -38,6 +40,67 @@ class Communicator
  */
 virtual void initialize() = 0;
 
+/**
+ * Define the start point of a collective communication.
+ *
+ * Note: nested start/stop pairs are not supported and will lead to undefined behavior.
+ */
+virtual void start() = 0;
+
+/**
+ * Blocked until all communication since the *start* has been completed.
+ */
+virtual void stop() = 0;
+
+/**
+ * Send data to a remote rank.
+ *
+ * @param[in] buf           Data buffer to send to remote rank
+ * @param[in] count         Number of elements to send
+ * @param[in] element_size  Size of each element
+ * @param[in] dest          Destination rank
+ */
+virtual void send(const void *buf, int64_t count, int element_size, int dest) = 0;
+
+/**
+ * Receive data from a remote rank.
+ *
+ * @param[in] buf           Receive buffer to place received data into
+ * @param[in] count         Number of elements to receive
+ * @param[in] element_size  Size of each element
+ * @param[in] source        Source rank
+ */
+virtual void recv(void *buf, int64_t count, int element_size, int source) = 0;
+
+/**
+ * Close the endpoints, free up used communication resources, and stop the communication runtime.
+ */
+virtual void finalize() = 0;
+
+int mpi_rank;
+int mpi_size;
+int current_device;
+cudaStream_t comm_stream;
```

Comment:
Worth adding comments that this is a stream reserved for communication-related kernels/copies, and that it's created/destroyed by the communicator internally. There might be value in adding a way to pass the stream as an argument from outside the communicator, in case the application wants to reuse streams for something else or enable easier pipelining? Say compressed kernels could use the same stream as communications (would that be a valid use case?)

Reply:
I moved `comm_stream` inside `NCCLCommunicator`, since currently only NCCL uses this stream.

> There might be value in adding a way to pass the stream as an argument from outside the communicator

I can see that for NCCL. This attribute is public so any class can access it.

For UCX, we have no control over which stream UCX is using.